[LV] Initial support for stores in early exit loops

huntergr-arm opened the pull request:
Adds some basic support for a simple early exit loop with a store.

This is vectorized such that when the next vector iteration would exit, we bail out to the scalar loop to handle the exit.

The main complication faced is transforming the vplan; it feels quite fragile at the moment, and might be better done up front at initial recipe creation time. Thoughts?

An alternative would be to use tail-folding, but that requires doing brka/brkb inside the loop, which could be costly. And would still likely require moving some IR, just not copying to the preheader.

(One advantage to this approach is that architectures without predication could potentially take advantage of ee autovec, though I suspect we'd want this driven by pragmas instead of just guessing about a possible trip count)

huntergr-arm committed:
[LV] Initial support for stores in early exit loops

Adds some basic support for a simple early exit loop with a store.

This is vectorized such that when the next vector iteration would
exit, we bail out to the scalar loop to handle the exit.

david-arm commented:
It's worth noting that brka/brkb is very much SVE-specific here, but there are other targets where it may be less of an issue.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1732,13 +1794,53 @@ bool LoopVectorizationLegality::isVectorizableEarlyExitLoop() {
+      SafetyInfo.computeLoopSafetyInfo(TheLoop);
+      // FIXME: We may have multiple levels of conditional loads, so will
+      //        need to improve on outright rejection at some point.
+      if (!SafetyInfo.isGuaranteedToExecute(*LI, DT, TheLoop)) {
```
Is this even possible at the moment? Given the branch condition comes from a icmp, which comes from a load, then the only way this might not execute is if the early exiting block is itself only executed conditionally. This isn't a loop structure we currently support. For example, a loop like this:

```
for.body:
   %cmp1 = icmp ...
   br i1 %cmp1, label %test.exit, %for.inc

test.exit:
   %ld1 = load i8, ...
   %cmp2 = icmp eq i8 %ld1, ...
   br i1 %cmp2, label %early.exit, label %for.inc

for.inc:
   ... branch back to for.loop or leave loop ...
```

huntergr-arm replied:
More future work, hence the FIXME.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1708,16 +1755,31 @@ bool LoopVectorizationLegality::isVectorizableEarlyExitLoop() {
+  bool HasStore = false;
   for (auto *BB : TheLoop->blocks())
     for (auto &I : *BB) {
+      if (StoreInst *SI = dyn_cast<StoreInst>(&I)) {
```
Does it matter if the store is prior to or after the early exit?

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1209,6 +1210,36 @@ bool LoopVectorizationLegality::canVectorizeMemory() {
+  //        be even if LAA thinks it is due to performing the load for the
+  //        vector iteration i+1 in vector iteration i.
+  if (isConditionCopyRequired()) {
+    assert(EarlyExitLoad.has_value() && "EE Store without condition load.");
```
If you change `isConditionCopyRequired` to test for `EarlyExitLoad` having a value, then you don't need this assert.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1209,6 +1210,36 @@ bool LoopVectorizationLegality::canVectorizeMemory() {
+  if (isConditionCopyRequired()) {
+    assert(EarlyExitLoad.has_value() && "EE Store without condition load.");
+
+    if (LAI->canVectorizeMemory()) {
```
Better to hoist this out and call it once to be reused here and below.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1209,6 +1210,36 @@ bool LoopVectorizationLegality::canVectorizeMemory() {
+          return false;
+        }
+      }
+    }
```
What happens in the else case? Looks like we fall through to:

```
  if (!LAI->canVectorizeMemory())
    return canVectorizeIndirectUnsafeDependences();
```

and so we may still treat the loop as legal. Worth adding a test for this case I think.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1209,6 +1210,36 @@ bool LoopVectorizationLegality::canVectorizeMemory() {
+      const auto *Deps = DepChecker.getDependences();
+
+      for (const MemoryDepChecker::Dependence &Dep : *Deps) {
+        if (Dep.getDestination(DepChecker) == EarlyExitLoad ||
```
Do we have to worry about other loads too? The header block could have several loads, with only one of them contributing to the branch condition.

david-arm reviewed (commented):
Thanks for this! I've only reviewed the legality bits, but thought I'd leave the comments I have so far.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1656,6 +1688,21 @@ bool LoopVectorizationLegality::isVectorizableEarlyExitLoop() {
+        if (Cmp && Cmp->hasOneUse() &&
+            TheLoop->isLoopInvariant(Cmp->getOperand(1))) {
+          LoadInst *Load = dyn_cast<LoadInst>(Cmp->getOperand(0));
+          if (Load && Load->hasOneUse() && TheLoop->contains(Load))
```
The store can also have a dependence with another load in the loop that doesn't contribute to the early exit condition.

huntergr-arm replied:
So there shouldn't be a problem with stores since dependencies for the EarlyExitLoad are checked later (in canVectorizeMemory()), but I didn't check for a uniform load. I've added the !loopInvariant check for now.

Technically it would have been rejected by the vplan transformation and the plan deleted (since that's looking for a load with an address based on the canonical IV), but it's nice if we can reject earlier instead.

huntergr-arm commented on llvm/include/llvm/Transforms/Vectorize/LoopVectorizationLegality.h:
```diff
@@ -407,6 +407,13 @@ class LoopVectorizationLegality {
+  bool isConditionCopyRequired() const { return RequiresEarlyExitConditionCopy; }
+
+  /// Returns the load instruction, if any, nearest to an uncountable early
+  /// exit.
```
The load closest in the IR graph to the exit (ideally, the load whose result is used directly in a comparison). I didn't spend much time on the naming ;)

huntergr-arm commented on llvm/include/llvm/Transforms/Vectorize/LoopVectorizationLegality.h:
```diff
@@ -407,6 +407,13 @@ class LoopVectorizationLegality {
   }
 
+  /// Returns true if this is an early exit loop containing a store.
+  bool isConditionCopyRequired() const { return RequiresEarlyExitConditionCopy; }
```
The store can be anywhere in the loop, since we either execute a full vector iteration as normal or we bail out to the scalar tail if an exit would occur in the middle.

We need to do this for all exits, so a single transformation to copy and reorder the exit condition IR is sufficient.

huntergr-arm commented on llvm/include/llvm/Transforms/Vectorize/LoopVectorizationLegality.h:
```diff
@@ -654,6 +661,16 @@ class LoopVectorizationLegality {
+  /// The load used to determine an uncountable early-exit condition. This is
+  /// only used to allow further analysis in canVectorizeMemory if we found
+  /// what looks like a valid early exit loop with store beforehand.
+  std::optional<LoadInst *> EarlyExitLoad;
```
This would fail the check that the second operand to the compare is loop invariant. There's a backup check in the vplan transform that would reject it at that point too.

In future, it would be nice to increase the number of cases handled, but I've tried to keep it simple for now and just come up with a list of extra cases to handle later.

Some of the possible future work:

1.  Supporting a combined condition; e.g. no escaping value, but two exit conditions or'd together. Needs work in ScalarEvolution? (This actually applies to the sample loop I chose; for the IR test I manually edited the IR to create 2 exits again)

2. Multiple uses of loads or comparisons in the exit IR chain; requires introducing a PHI node for current vector iteration value.

3. Supporting a chain of conditional loads; e.g. early exit is itself in a conditional block so might not always execute.

4. Supporting a non-const second term for comparison, or more varied comparisons (e.g. bit test instead of icmp).

5. Tail folding of EE loops with a store.

6. Dynamic bounds for known-dereferenceable load (via e.g. call void @llvm.assume(i1 true) [ "align"(ptr %pred, i64 2), "dereferenceable"(ptr %pred, i32 %n_bytes) ])

huntergr-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1656,6 +1688,21 @@ bool LoopVectorizationLegality::isVectorizableEarlyExitLoop() {
+      if (BranchInst *Br = dyn_cast<BranchInst>(BB->getTerminator())) {
+        // FIXME: Handle exit conditions with multiple users, more complex exit
+        //        conditions than br(icmp(load, loop_inv)).
+        ICmpInst *Cmp = dyn_cast<ICmpInst>(Br->getCondition());
```
It should, but this is supposed to be a minimal implementation :)

huntergr-arm committed:
Remove redundant variable, add helper to clear ee state, change load check to !invariant

huntergr-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1732,13 +1794,53 @@ bool LoopVectorizationLegality::isVectorizableEarlyExitLoop() {
     return false;
+  } else {
+    // Read-only loop.
+    // FIXME: as with the loops with stores, only the loads contributing to
```
No, that's why it's a FIXME. Future work :)

huntergr-arm committed:
Move exit load checks after store detection

huntergr-arm committed:
Improve memory vectorization checks

huntergr-arm commented:
I've started using vplan pattern matchers in the transform; still feels like there's potential for it to exclude a bunch of valid loops if something is slightly different.

At this point I think I'll precommit the tests and separate out some parts into additional PRs.

huntergr-arm commented:
Hi @annamthomas,

I've been working on landing pieces of this work as individual PRs. As you note, the legality checks have been committed, as has #152530 to identify the vplan recipes contributing to an uncountable exit.

The latter was split from #148626, which is now my focus and is currently awaiting review. That patch has the core transform to handle uncountable exits that would occur in the _next_ vector iteration in the scalar loop instead, which was the plan I mentioned in the RFC.

A rough idea of the resulting code would be:

```
;; after SCEV/min.iter checks
vector.ph:
  ;; Perform load for uncountable condition for _first_ vector iteration
  ;; Compare and reduce result of load, and either branch to the scalar
  ;; preheader if there would be an exit, or vector.ph.split otherwise
vector.ph.split:
  br vector.body
vector.body:
  iv = phi....
  ;; Move iv increment operation early enough for the load
  iv.next = add iv, step
  ;; Load uncountable exit data for the _next_ vector iteration
  addr = gep uncountable_base, iv.next
  uc.load = load addr
  ;; compare and reduce result of load, combine with countable
  ;; exit condition
  ;; branch back to vector body or middle block
middle.block:
  ;; Determine whether we exited due to reaching the loop upper bound
  ;; if not, branch to the scalar.ph to handle either type of exit
```

This doesn't currently support using the uncountable exit condition load value for anything beyond the exit; that requires a small follow-up to introduce a phi if there are additional uses.

After that lands, I can rebase this patch on top of it. There shouldn't be much left to do at that point for the minimum supported case, and extending the set of loops handled can be dealt with in additional PRs.

Hopefully that explains enough? Please let me know if something's not clear.

annamthomas commented:
thanks @huntergr-arm , I have a high level question - there is an additional constraint that the store has to happen *before* the early-exit. Is this fundamental to the implementation? How hard is it to extend to support stores anywhere in the loop?

I think it's important to also discuss how these loops can be supported. For example, we have predication support in many architectures. If we generate predicates for all instructions in the loop (using the early-exit form), then I think it wouldn't matter whether the operation occurs before or after the early-exit. We had a proposal a while ago about this: https://github.com/annamthomas/work-notes/blob/main/MultiDataDependentExitVectorization.rst#full-predication-approach (it also allows supporting for multiple early-exits from the loop: pls only look at the full-predication-approach). 

Also, @david-arm,  I noticed this constraint exists in the read-only loop as well (and couple of other constraints such as there is no in-built predication in the loop, i.e. early exit should feed directly into the latch). 

In short, the main question is if the implementation allows for extendability to support common patterns such as:
1. stores and loads *anywhere* within the loop.
2. existing predication within the loop
3. multiple early-exits 

I believe this is relevant for the already existing logic for read-only early exit loops. (@david-arm  I don't want to hijack this PR, but we are seeing some cases downstream where supporting multiple early-exits in read-only loops is beneficial. That will need updating the current cost model, but thats a separate topic).

huntergr-arm commented:
Hi @annamthomas,

No, there's nothing fundamental about where stores (or loads) are in the loop for my work; that's why my initial approach determines whether the next vector iteration would have an uncountable exit. Loads or stores can occur anywhere in the loop, and we bail out to the scalar loop if we know we're exiting early. One of my tests in early_exit_store_legality.ll (`@loop_contains_store_in_latch_block`) is intended to cover this. I'm not sure if this patch will be sufficient when it lands, but the plan is definitely to vectorize such loops.

After the initial work lands, there will be some additional work to enable the same to function with tail-folding. For SVE at least, we have the brkb and brka instructions to create masks for operations before and after a given exit so that we do not execute memory operations to locations that the scalar loop would not touch (or anything else which might fault).
